art/attacks/evasion/imperceptible_asr/imperceptible_asr.py

```diff
@@ -0,0 +1,621 @@
+        imperceptible to the human ear.
+
+        :param x: An array with the original inputs to be attacked.
+        :param x_adversarial: An array with the adversarial examples.
+        :param y: Target values of shape (batch_size,). Each sample in `y` is a string and it may possess different
+            lengths. A possible example of `y` could be: `y = np.array(['SIXTY ONE', 'HELLO'])`.
+        :return: An array with the imperceptible, adversarial outputs.
+        """
+        batch_size = x.shape[0]
+        alpha_min = 0.0005
+
+        alpha = np.array([self.alpha] * batch_size)
+        loss_theta_previous = [np.inf] * batch_size
+        x_imperceptible = [None] * batch_size
+        # if inputs are *not* ragged, we can't multiply alpha * gradients_theta
+        if x.ndim != 1:
+            alpha = np.expand_dims(alpha, axis=-1)
+
+        perturbation = x_adversarial - x
+        x_perturbed = x_adversarial.copy()
+
+        for i in range(self.max_iter_2):
+            # get loss gradients of both losses
+            gradients_net = self.estimator.loss_gradient(x_perturbed, y, batch_mode=True)
+            gradients_theta, loss_theta = self._loss_gradient_masking_threshold(perturbation, x)
+
+            # perform gradient descent steps
+            x_perturbed = x_perturbed - self.learning_rate_2 * (gradients_net + alpha * gradients_theta)
+
+            if i % 20 == 0 or i % 50 == 0:
+                prediction = self.estimator.predict(x_perturbed, batch_size=batch_size)
+                for j in range(batch_size):
+                    # validate if adversarial target succeeds, i.e. f(x_perturbed)=y
+                    if i % 20 == 0 and prediction[j] == y[j].upper():
+                        # increase alpha
+                        alpha[j] *= 1.2
+                        # save current best imperceptible, adversarial example
+                        if loss_theta[j] < loss_theta_previous[j]:
+                            x_imperceptible[j] = x_perturbed[j]
+                            loss_theta_previous[j] = loss_theta[j]
+
+                    # validate if adversarial target fails, i.e. f(x_perturbed)!=y
+                    if i % 50 == 0 and prediction[j] != y[j].upper():
+                        # decrease alpha
+                        alpha[j] = max(alpha[j] * 0.8, alpha_min)
+                logger.info("Current iteration %s, alpha %s, loss theta %s", i, alpha, loss_theta)
+
+        # return perturbed x if no adversarial example found
+        for j in range(batch_size):
+            if x_imperceptible[j] is None:
+                logger.critical("Adversarial attack stage 2 for x_%s was not successful", j)
+                x_imperceptible[j] = x_perturbed[j]
+
+        return np.array(x_imperceptible, dtype=object)
+
+    def _loss_gradient_masking_threshold(self, perturbation: np.ndarray, x: np.ndarray) -> np.ndarray:
```

Comment:
I think we need to add the return type of the second return value.

Reply:
Yes, you are right! Nice catch :+1: